File: go/cmd/cadreen/commands/escalations.go
package commands

import (
	"encoding/json"
	"fmt"

	"github.com/timothy-billingrails/cadreen-sdks/go/cmd/cadreen/output"

	cadreen "github.com/timothy-billingrails/cadreen-sdks/go/cadreen"

	"github.com/spf13/cobra"
)

var escalationsCmd = &cobra.Command{
	Use:   "escalations",
	Short: "Manage escalations (list, get, resolve)",
	Long:  `View and resolve escalations that need human attention.`,
}

var escalationsListCmd = &cobra.Command{
	Use:   "list",
	Short: "List escalations",
	RunE: func(cmd *cobra.Command, args []string) error {
		requireAuth()

		resp, err := rawDo("GET", "/api/v1/cadreen/escalations", nil)
		if err != nil {
			return handleAPIError(err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != 200 {
			return handleHTTPError(resp)
		}

		var escalations cadreen.ListEscalationsResponse
		if err := json.NewDecoder(resp.Body).Decode(&escalations); err != nil {
			return err
		}

		format := outputFormat()
		switch format {
		case output.FormatJSON:
			output.Print(escalations, format)
		case output.FormatVerbose:
			printEscalationsVerbose(escalations.Escalations)
		default:
			printEscalationsText(escalations.Escalations)
		}

		return nil
	},
}

var escalationsGetCmd = &cobra.Command{
	Use:   "get [id]",
	Short: "Get an escalation by ID",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		requireAuth()

		resp, err := rawDo("GET", "/api/v1/cadreen/escalations/"+args[0], nil)
		if err != nil {
			return handleAPIError(err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != 200 {
			return handleHTTPError(resp)
		}

		var esc cadreen.Escalation
		if err := json.NewDecoder(resp.Body).Decode(&esc); err != nil {
			return err
		}

		format := outputFormat()
		switch format {
		case output.FormatJSON:
			output.Print(esc, format)
		default:
			printEscalationDetail(esc)
		}

		return nil
	},
}

var escalationsResolveCmd = &cobra.Command{
	Use:   "resolve [id] [resolution]",
	Short: "Resolve an escalation",
	Long: `Resolve an escalation with a resolution message.

Examples:
  cadreen escalations resolve esc_123 "Approved refund of $50"
  cadreen escalations resolve esc_456 "Denied — exceeds policy limit"`,
	Args: cobra.ExactArgs(2),
	RunE: func(cmd *cobra.Command, args []string) error {
		requireAuth()

		client := newClient()
		esc, err := client.ResolveEscalation(cmd.Context(), args[0], args[1])
		if err != nil {
			return handleAPIError(err)
		}

		format := outputFormat()
		if format == output.FormatJSON {
			output.Print(esc, format)
		} else {
			fmt.Printf("Escalation %s resolved.\n", esc.ID)
		}

		return nil
	},
}

func init() {
	escalationsCmd.AddCommand(escalationsListCmd)
	escalationsCmd.AddCommand(escalationsGetCmd)
	escalationsCmd.AddCommand(escalationsResolveCmd)
	rootCmd.AddCommand(escalationsCmd)
}

func printEscalationsText(escs []cadreen.Escalation) {
	if len(escs) == 0 {
		fmt.Println("No escalations.")
		return
	}

	fmt.Printf("Escalations (%d):\n\n", len(escs))
	for _, e := range escs {
		statusIcon := "!"
		if e.Status == "resolved" {
			statusIcon = "✓"
		}
		fmt.Printf("  %s %s\n", statusIcon, e.ID)
		if e.Intent != "" {
			fmt.Printf("    %s\n", e.Intent)
		}
		if e.Severity != "" {
			fmt.Printf("    Severity: %s\n", e.Severity)
		}
		fmt.Println()
	}
}

func printEscalationsVerbose(escs []cadreen.Escalation) {
	for _, e := range escs {
		printEscalationDetail(e)
	}
}

func printEscalationDetail(e cadreen.Escalation) {
	fmt.Printf("ID:       %s\n", e.ID)
	fmt.Printf("Status:   %s\n", e.Status)
	if e.Intent != "" {
		fmt.Printf("Intent:   %s\n", e.Intent)
	}
	if e.Severity != "" {
		fmt.Printf("Severity: %s\n", e.Severity)
	}
	if e.ErrorMessage != "" {
		fmt.Printf("Error:    %s\n", e.ErrorMessage)
	}
	if e.HumanPrompt != "" {
		fmt.Printf("Prompt:   %s\n", e.HumanPrompt)
	}
	if e.Resolution != "" {
		fmt.Printf("Resolution: %s\n", e.Resolution)
	}
	fmt.Println()
}

File: python/cadreen/resources/escalations.py
from __future__ import annotations

from typing import Any

from ..client import HttpClient
from ..types import (
    ListEscalationsResponse,
    Escalation,
    Pagination,
)


class EscalationsResource:
    def __init__(self, client: HttpClient) -> None:
        self._client = client

    async def list(self) -> ListEscalationsResponse:
        raw = await self._client.get("/api/v1/cadreen/escalations")
        escalations = [
            Escalation(
                id=e["id"],
                status=e["status"],
                intent=e.get("intent"),
                category=e.get("category"),
                execution_id=e.get("execution_id"),
                tool_name=e.get("tool_name"),
                error_message=e.get("error_message"),
                severity=e.get("severity"),
                human_prompt=e.get("human_prompt"),
                suggestions=e.get("suggestions"),
                created_at=e.get("created_at"),
                resolved_at=e.get("resolved_at"),
                resolved_by=e.get("resolved_by"),
                resolution=e.get("resolution"),
            )
            for e in raw.get("escalations", [])
        ]
        pagination = None
        if raw.get("pagination"):
            p = raw["pagination"]
            pagination = Pagination(limit=p["limit"], offset=p["offset"], has_more=p["has_more"])
        return ListEscalationsResponse(
            escalations=escalations,
            count=raw.get("count", 0),
            pagination=pagination,
        )

    async def get(self, id: str) -> Escalation:
        raw = await self._client.get(f"/api/v1/cadreen/escalations/{id}")
        return Escalation(
            id=raw["id"],
            status=raw["status"],
            intent=raw.get("intent"),
            category=raw.get("category"),
            execution_id=raw.get("execution_id"),
            tool_name=raw.get("tool_name"),
            error_message=raw.get("error_message"),
            severity=raw.get("severity"),
            human_prompt=raw.get("human_prompt"),
            suggestions=raw.get("suggestions"),
            created_at=raw.get("created_at"),
            resolved_at=raw.get("resolved_at"),
            resolved_by=raw.get("resolved_by"),
            resolution=raw.get("resolution"),
        )

    async def resolve(self, id: str, decision: str) -> Escalation:
        raw = await self._client.post(
            f"/api/v1/cadreen/escalations/{id}/resolve",
            {"decision": decision},
        )
        return Escalation(
            id=raw["id"],
            status=raw["status"],
            intent=raw.get("intent"),
            category=raw.get("category"),
            execution_id=raw.get("execution_id"),
            tool_name=raw.get("tool_name"),
            error_message=raw.get("error_message"),
            severity=raw.get("severity"),
            human_prompt=raw.get("human_prompt"),
            suggestions=raw.get("suggestions"),
            created_at=raw.get("created_at"),
            resolved_at=raw.get("resolved_at"),
            resolved_by=raw.get("resolved_by"),
            resolution=raw.get("resolution"),
        )
